move the add/edit description button to the app-bar

mawinter69 opened the pull request:
So far the add/edit description button is rendered on its own line to the far right. This leads to additional space between the description or the title and the following content.

This change moves the add/edit description button to the app-bar where applicable (jobs, runs, computers, users and labels). For views it stays as it is as they don't have a title/app-bar. One also can't just allow to change the description via the configure page as the `All` View has no configure page.

The change is backwards compatible so plugins using the editableDescription taglib will work as before.

Requires changes in Matrix-project and maven-project plugins to align the new layout


### Before/after:
#### Job:
Before:
![image](https://github.com/jenkinsci/jenkins/assets/17767050/b0699926-6c89-43c9-8af2-96a51c4833af)
After:
![image](https://github.com/jenkinsci/jenkins/assets/17767050/d1d06459-0d83-4762-8ab1-26661dc3b291)
A pipeline job:
![image](https://github.com/jenkinsci/jenkins/assets/17767050/9cf93d8b-c338-4ca9-b73c-73057188856e)


#### Run:
Before:
![image](https://github.com/jenkinsci/jenkins/assets/17767050/76f43e72-6502-4f77-a721-1f02efa3f5cd)
After:
![image](https://github.com/jenkinsci/jenkins/assets/17767050/1bf7768a-a828-44e3-8c48-67556d2e2668)
After clicking the `Add Description` button
![image](https://github.com/jenkinsci/jenkins/assets/17767050/c03155be-c73a-4f24-9626-96e5db1703c9)




#### User:
Before:
![image](https://github.com/jenkinsci/jenkins/assets/17767050/f75bf96a-57f7-45b9-bc23-0bf22c2e894a)
After:
![image](https://github.com/jenkinsci/jenkins/assets/17767050/acd33f02-a2aa-4fdd-bda2-d9efe2ca3946)

#### Computer:
Before:
![image](https://github.com/jenkinsci/jenkins/assets/17767050/bd159b1d-19e5-4a88-b465-dc42fa60c220)
After:
![image](https://github.com/jenkinsci/jenkins/assets/17767050/8e28d889-9cf8-473f-a1eb-36cd46339de0)

#### Label:
Before:
![image](https://github.com/jenkinsci/jenkins/assets/17767050/f57e79a9-69d1-4694-99a2-373f5873904c)
After:
![image](https://github.com/jenkinsci/jenkins/assets/17767050/660eeb0c-9b71-4d26-9cf7-a1d4a6d664d1)





<!-- Comment:
A great PR typically begins with the line below.
Replace XXXXX with the numeric part of the issue ID you created in Jira.
Note that if you want your changes backported into LTS, you need to create a Jira issue. See https://www.jenkins.io/download/lts/#backporting-process for more information.
-->


<!-- Comment:
If the issue is not fully described in Jira, add more information here (justification, pull request links, etc.).

 * We do not require Jira issues for minor improvements.
 * Bug fixes should have a Jira issue to facilitate the backporting process.
 * Major new features should have a Jira issue.
-->

### Testing done
Manual Testing
<!-- Comment:
Provide a clear description of how this change was tested.
At minimum this should include proof that a computer has executed the changed lines.
Ideally this should include an automated test or an explanation as to why this change has no tests.
Note that automated test coverage is less than complete, so a successful PR build does not necessarily imply that a computer has executed the changed lines.
If automated test coverage does not exist for the lines you are changing, you must describe the scenario(s) in which you manually tested the change.
For frontend changes, include screenshots of the relevant page(s) before and after the change.
For refactoring and code cleanup changes, exercise the code before and after the change and verify the behavior remains the same.
-->

### Proposed changelog entries

- Move Add/Edit description to app bar.

<!-- Comment:
The changelog entry should be in the imperative mood; e.g., write "do this"/"return that" rather than "does this"/"returns that".
For examples, see: https://www.jenkins.io/changelog/

Remove JENKINS-XXXXX if there is no issue for the pull request.

You may add multiple changelog entries if applicable by adding a new entry to the list, e.g.
- JENKINS-123456, First changelog entry
- Second changelog entry
-->

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/core-pr-reviewers.
-->

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

mawinter69 committed:
move the add/edit description button to the app-bar

So far rhe add/edit description button is rendered as button on its own
line to the far right. This leads to additional space the description or
the title and the following content.

This change moves the add/edit description button to the app-bar where
applicable (jobs, runs, computers, users). For views it stays as it is
as they don't have title/app-bar. One also can't just allow to change
the description via the configure page as the All View has no configure
page.

mawinter69 commented:
Some questions to think about
- should I directly create an overflow button that takes add/edit description

On the run page
- also move the keep log forever to overflow
- move the info about when the build started and how long it took to the left like e.g. the info about the cause.

![image](https://github.com/jenkinsci/jenkins/assets/17767050/e580e2c1-d7b8-412e-85e6-01704868440a)

mawinter69 commented:
@jenkinsci/sig-ux

janfaracik commented:
Like it 👍

> * make the add/edit description button a secondary button?

Seems sensible yeah.

> * should I directly create an overflow button that takes add/edit description?

Could do, I do wonder how much benefit/use the 'Add description' button gets to warrant it existing - rather than just existing in the 'Configure' page.

> * also move the keep build forever to overflow?

Sounds good.

> * move the info about when the build started and how long it took to the left like e.g. the info about the cause.

Want to eventually move this section into its own extendable card (in a similar fashion to https://jenkins-redesign.vercel.app/project/build/375). I don't think the current summary components are the best use of space.

mawinter69 committed:
Merge branch 'master' into editable-description

mawinter69 committed:
make secondary button

yaroslavafenkin commented on core/src/main/resources/hudson/model/Job/index.jelly:
```diff
@@ -40,6 +40,9 @@ THE SOFTWARE.
           </h1>
         </div>
+        <div class="jenkins-app-bar__controls">
+          <t:editDescriptionButton permission="${it.UPDATE}"/>
```
There's no `Job.UPDATE`, is there? Is there any reason you went for a change?

mawinter69 replied:
thanks fixed

yaroslavafenkin commented on war/src/main/webapp/scripts/hudson-behavior.js:
```diff
@@ -1901,8 +1901,11 @@ function xor(a, b) {
   var d = document.getElementById("description");
-  d.firstElementChild.nextElementSibling.innerHTML =
-    "<div class='jenkins-spinner'></div>";
+  let button = d.firstElementChild.nextElementSibling;
```
```
hudson-behavior.js:1904 Uncaught TypeError: Cannot read properties of null (reading 'firstElementChild')
    at replaceDescription (hudson-behavior.js:1904:18)
    at HTMLAnchorElement.<anonymous> (editable-description.js:10:14)
```
when I hit edit description at `/manage/computer/(built-in)/`

mawinter69 replied:
ahh the built-in has no description. Worked around by only showing the button when there is an actual div with id `description`

mawinter69 commented:
> Like it 👍
> 
> > * make the add/edit description button a secondary button?
> 
> Seems sensible yeah.
> 
> > * should I directly create an overflow button that takes add/edit description?
> 
> Could do, I do wonder how much benefit/use the 'Add description' button gets to warrant it existing - rather than just existing in the 'Configure' page.
> 
> > * also move the keep build forever to overflow?
> 
> Sounds good.
> 
> > * move the info about when the build started and how long it took to the left like e.g. the info about the cause.
> 
> Want to eventually move this section into its own extendable card (in a similar fashion to https://jenkins-redesign.vercel.app/project/build/375). I don't think the current summary components are the best use of space.

For now I only made the description button a secondary button.

mawinter69 committed:
feedback

- fix permission in job page
- don't show the button for built-in node
- improve jelly doc

Kevin-CB commented:
LGTM security wise

timja reviewed (approved):
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!
